### PR TITLE
SSCSSI-98-21: C4: hearingBooked - ProcessHmcMessageService (Only cove…

### PR DESCRIPTION
**Pull request by srochani**
…rs List Assist Hearing booking)

### Jira link

See [PROJ-XXXXXX](https://tools.hmcts.net/jira/browse/PROJ-XXXXXX)

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import uk.gov.hmcts.reform.sscs.model.hmc.message.HmcMessage;
+import uk.gov.hmcts.reform.sscs.model.hmc.reference.HmcStatus;
+import uk.gov.hmcts.reform.sscs.model.single.hearing.HearingGetResponse;
+import uk.gov.hmcts.reform.sscs.service.CcdCaseService;
+import uk.gov.hmcts.reform.sscs.service.HmcHearingApiService;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ProcessHmcMessageServiceV2 {
```
Comment (SomeshRA) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java:32]: Add a new interface with "processEventMessage" method
Add - @ConditionalOnProperty(name = "${feature.process-event-message-v2.enabled}", havingValue = "true")

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/HmcHearingsEventTopicListener.java [hunk 2]
```diff
@@ -28,21 +31,26 @@ public class HmcHearingsEventTopicListener {
...
+    @Value("${feature.process-event-message-v2.enabled}")
+    private boolean processEventMessageV2Enabled;
     @Value("${feature.bypass-hearing-api-service.enabled}")
     private boolean isByPassHearingServiceEnabled;
 
     private static final String HMCTS_DEPLOYMENT_ID = "hmctsDeploymentId";
 
     public HmcHearingsEventTopicListener(@Value("${sscs.serviceCode}") String sscsServiceCode,
-                                         ProcessHmcMessageService processHmcMessageService) {
+                                         ProcessHmcMessageService processHmcMessageService,
```
Comment (SomeshRA) [inline src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/HmcHearingsEventTopicListener.java:49]: Autowire new interface - which inject either ProcessHmcMessageService/ProcessHmcMessageServiceV2 based on "feature.process-event-message-v2.enabled"

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/HmcHearingsEventTopicListener.java [hunk 3]
```diff
@@ -72,18 +80,24 @@ public void onMessage(JmsBytesMessage message) throws JMSException, HmcEventProc
...
-                        caseId,
-                        hearingId
+                    "Attempting to process message from HMC hearings topic for event {}, Case ID {}, and Hearing ID {}.",
+                    hmcMessage.getHearingUpdate().getHmcStatus(),
+                    caseId,
+                    hearingId
                 );
 
-                processHmcMessageService.processEventMessage(hmcMessage);
+                if (processEventMessageV2Enabled) {
```
Comment (SomeshRA) [inline src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/HmcHearingsEventTopicListener.java:89]: So this condition evaluation will not be required based on above change.

File: src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+            hearingUpdateService.setWorkBasketFields(hearingId, sscsCaseData, hmcMessageStatus);
+
+            return resolveEventType(sscsCaseData, hmcMessageStatus, hearingResponse, hearingId);
+
+        };
+
+        SscsCaseDetails initialSscsCaseDetails = ccdCaseService.getCaseDetails(caseId);
+        IdamTokens idamTokens = idamService.getIdamTokens();
+        DynamicEventUpdateResult dynamicEventUpdateResult = mutator.apply(initialSscsCaseDetails);
```
Comment (SomeshRA) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java]: I suppose this is only invoked to get event type and whether to commit. While the actual case data update is done in the updateCaseV2DynamicEvent. Hence, if so than just invoking resolveEventType(initialSscsCaseDetails.getData(), hmcMessageStatus, hearingResponse, hearingId) should suffice.

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java [hunk 5]
```diff
@@ -52,7 +52,7 @@ public void handleIncomingMessage(HearingRequest message) throws TribunalsEventP
         HearingState event = message.getHearingState();
 
         log.info("Attempting to process hearing event {} from hearings event queue for case ID {}",
-                event, caseId);
+                 event, caseId);
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java:55]: is it easy for you to get rid off all the reformatting-only changes.... it just clutters up the PR

File: src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        updateCcdCaseService.updateCaseV2DynamicEvent(caseId, idamTokens, mutator);
+
+        log.info(
+            "Hearing message {} processed using V2 for case reference {}",
+            hmcMessage.getHearingId(),
+            hmcMessage.getCaseId()
+        );
+    }
+
+    @Override
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java:101]: I thought you were gonna add the ConditionalOnProperty value to the class.... is there a reason you decided not to do that?

File: src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceFactory.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package uk.gov.hmcts.reform.sscs.service.hmc.topic;
+
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+public class ProcessHmcMessageServiceFactory {
```
Comment (SomeshRA) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceFactory.java:10]: I don't think we need this factory as only instance of the service will be enabled based on the property.

File: src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java (hunk 1 above)
Comment (srochani) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/HmcHearingsEventTopicListener.java (hunk 2 above)
Comment (srochani) [reply src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/HmcHearingsEventTopicListener.java]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/HmcHearingsEventTopicListener.java (hunk 3 above)
Comment (srochani) [reply src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/HmcHearingsEventTopicListener.java]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java (hunk 4 above)
Comment (srochani) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java (hunk 5 above)
Comment (srochani) [reply src/main/java/uk/gov/hmcts/reform/sscs/jms/listener/TribunalsHearingsEventQueueListener.java]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java (hunk 6 above)
Comment (srochani) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceV2.java]: done

File: src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceFactory.java (hunk 7 above)
Comment (srochani) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/hmc/topic/ProcessHmcMessageServiceFactory.java]: done

**Review by sanjayparekh-hmcts — DISMISSED** [review]

**Review by SomeshRA — DISMISSED** [review]
LGTM

**Review by sanjayparekh-hmcts — APPROVED** [review]

**Review by SomeshRA — APPROVED** [review]

**Pull request merged**